Set Language Attribute of the `.pages` Element

Gazook89 opened the pull request:
This pr resolves #1343 by adding a "Language" property in the properties editor.   The input value is then sent to the BrewRenderer component and inserted into the `.pages` element.

The property input field is a datalist that shows the language code + the name of the language in that language.  Adding languages to this list is easy, just add the code and the "name" is automatically added.  There is a very simple pattern validation, and an error message that displays under the input field....this isn't my favorite, since the text is so small and it's on a gray background (so it is difficult to get high contrast), but this can be changed to match whatever is decided on PR #2374 so i didn't spend a lot of time on it.   Here is a picture of it now:

<img width="478" alt="image" src="https://user-images.githubusercontent.com/58999374/197408880-d4888223-f0d5-4f48-b253-25b9d9389320.png">

The pattern itself could be made significantly more robust, or we could use a new library to actually check to see if the language tag is valid (mentioned in attached Issue).

ericscheid commented:
Perhaps add a link to a resource that explains (in friendly terms) what language codes exist?

Perhaps laid out like 
`Language:   [_________________] _Language codes?_`

ericscheid commented on client/homebrew/editor/metadataEditor/metadataEditor.jsx:
```diff
@@ -225,6 +251,32 @@ const MetadataEditor = createClass({
+			<input type='text' className='value'
+				value={this.props.metadata.lang}
+				onChange={(e)=>this.handleFieldChange('lang', e)}
+				list='languageList'
```
Anytime an `input` has a `list` attribute, it is usually a good idea to turn off `autocomplete` .. otherwise the UI for autocomplete can also appear, and the input now has two drop-down lists of suggestions. It still works, but it looks very silly.

Gazook89 replied:
committed and pushed

Gazook89 commented:
One issue I can foresee is numerous versions of the question "why doesn't changing the language in properties translate my document to that language?".   I'm not sure how to avoid that...maybe changing the label to "Document Language"?   Or maybe we just tackle it if/when it becomes a problem.

Gazook89 commented:
Things left to do:

- [ ] currently the validation pattern just shows an error, but still accepts the value (and saves it to the document).  Likely it should only accept valid patterns.
- [ ] correctly set the default to 'en'
- [ ] confirm choices for the datalist.

calculuschild commented on server/homebrew.model.js:
```diff
@@ -15,6 +15,7 @@ const HomebrewSchema = mongoose.Schema({
 	description : { type: String, default: '' },
 	tags        : [String],
 	systems     : [String],
+	lang        : { type: String, default: '' },
```
should the default be "en"?

Gazook89 replied:
updated.

Gazook89 commented:
This is ready for review I think.

Right now:
1. When creating 'New' document, it defaults the Language input and the `.pages` element to `en`.
2. If you change the input, it updates the `.pages` element and also saves correctly to the `METAKEY` in localStorage.
3. So if you reload the page (or leave and return to the properties editor), it keeps the input value.
4. If you delete the value to an empty string, the `.pages` element will revert the lang attribute to `en`, and the next time you reopen properties editor it will show `en`.
5. Everything on the Edit page works pretty much the same way.

Before merging, might want to confirm the following:
1. What goes in the 'suggested' languages datalist?
2. Are we satisfied with the current minimal pattern?  Basically just requires 2-3 letters, and optionally followed by `-anything`.     Or, I just put together this regex which I think covers the ['common pattern' given on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang#language_tag_syntax):

```js
/^[a-zA-Z]{2,3}(-[A-Z]{1}[a-z]{3})?(-(?>[0-9]{3}|[A-Z]{2}))?$/gm
```
[*regex101.com*](https://regex101.com/r/5uqXQ2/1)

3. Finally, I'm just not sure about where to put a helpful link to the BCP 47 / language code reference list.  It could go in small text below the input....?  There could be little 'info-circle' buttons that open tooltips....?  I'm sort of the opinion that having a datalist of the most common languages is going to be enough until proven otherwise.

Gazook89 commented:
i updated the regex to the slightly more strict version I posted above (with a minor change, which I edited that previous comment to reflect).  And I added a couple more language tags suggested by Eric.  Finally, I fixed the autocomplete problem.

calculuschild commented:
With `en` as the default, the dropdown only lists English and Swedish as options:

![image](https://user-images.githubusercontent.com/5878534/203124849-4903beb3-6bfd-42b0-8f5a-cbf7b4d7e670.png)

Similarly, when any other option is selected, the dropdown lists only the current selection:

![image](https://user-images.githubusercontent.com/5878534/203124993-9e048cf3-db8d-4e81-aa9f-cca7cec007b4.png)

Looks like it's doing more of a "match" (svenska and english both have "en") rather than a selection box. Was very confusing to me, thinking: "why are we only providing English and Swedish as options?"


This has another curious interaction, where typing the first letter of a language triggers both the "autocomplete" match *and* the validator:

![image](https://user-images.githubusercontent.com/5878534/203126152-bd68179f-75c1-4213-a6c5-fa141c688b00.png)


Last, the dropdown is in a different position when the field is blank. Wherever it ends up, I prefer the dropdown in a consistent location:

![image](https://user-images.githubusercontent.com/5878534/203126379-1e6a7f0f-4147-4333-b790-0d5bdbc38458.png)

Gazook89 commented:
> Looks like it's doing more of an "autocomplete match" 

Yeah, it doesn't look like there is method to display all suggested options in datalist.  Datalist is conceptually a part of the autocomplete tag.  I did a bit of searching for how to display all options while still using datalist and didn't see anything, but there is probably a way to do it with js.  Or will just have to go with a js combobox. 

>  I prefer the dropdown in a consistent location:

Again, [after some searching](https://stackoverflow.com/questions/13693482/is-there-a-way-to-apply-a-css-style-on-html5-datalist-options), it doesn't appear that datalists can be given any css styling, so I can't pick a standard position.  There are libraries out there that provide custom inputs, or could just make our own.

> This has another curious interaction, where typing the first letter of a language triggers both the "autocomplete" match and the validator:

Would this bother you if they didn't overlap?  If so, making our own combobox and setting it to appear to the right would be one remedy.   Otherwise, would need to figure out debounce on the field validation.

I have another PR I just started in which I want a combobox, so maybe that's just what needs to be done.

ericscheid commented:
1.) The html `autocomplete` thing is a different thing from the `datalist` thing. You generally don't want both active on an input, mainly because of the clashing/competing UIs. (I think I might have mentioned this when playing with datalist for tags).

- `autocomplete` is a mechanism for automatically filling a specific value from either frequent usage or from a user profile, possibly offering a choice of same
- `datalist` is a mechanism for providing an extensive list of possible values for an input

One is saying ″here are recommendations specific to you″ and the other is saying "here is the full menu of options". Doing both simultaneously is incoherent.

(BTW, is `autoComplete` a React thing? The html attribute name is `autocomplete`.)

2.) For the validation thing you might want try delaying validation via a debounce so it only kicks in when the user stops typing. Alternatively, allow `[a-z]` while the input has focus, but use the fuller regex on blur.

3.) As to defaulting `en` .. maybe the field should be empty but with a placeholder of simply `en` (i.e. not a narrative description, a literal example value). We can leave the field blank when saving, as the renderer will insert `en` if blank. The datalist itself demonstrates a range of values so we don't need an extended list of examples in the placeholder.

4.) If at all possible, could the input element be shrunk to be closer in width to what the normal range of values will likely be. I'm not sure what CSS would be needed since there's `flex` in play here. Maybe `.metadataEditor .field.language>.value { flex: unset; width: 150px; }`. Doing this means the datalist will be physically constrained as to where it could appear.

calculuschild commented:
>  Or will just have to go with a js combobox.

I would probably prefer this anyway, since we already have custom dropdown menus implemented in multiple places.

>  I can't pick a standard position. There are libraries out there that provide custom inputs, or could just make our own.

Let's just make use of the dropdowns we already have (see the Themes selector). Making it text-editable should be trivial, and then we can have a more consistent style as well rather than suddenly using datalists in one place.

> Would this bother you if they didn't overlap?

That would help, but I would prefer they just don't appear at the same time if possible. If we are being shown an auto-complete suggestion, we already know the input isn't valid. We don't need both at once.

>Otherwise, would need to figure out debounce on the field validation.

We should probably do this anyway (different PR maybe?). Having an error show up immediately when I'm still typing is a little bit aggressive.

calculuschild commented:
> (BTW, is `autoComplete` a React thing? The html attribute name is `autocomplete`.)

Since React is using "JSX" rather than HTML, a few properties have slightly different spellings/capitalizations to not clash with existing Javascript names during executing (such as `className` vs `class` for CSS styling) . But when rendered to the DOM it does get converted to the standard HTML attribute.

Gazook89 commented:
> Edit: Ehhh.... maybe the datalist is fine. I just don't like that it hides options when you click on it and there is something in the box already. The autosuggest behavior of the datalist can stay (but turn autocomplete=off as ericsheid says, since that's a different thing). But I sure would like On Click to show the full list like a normal select box instead of limited to matching the current value.

I would imagine that onClick would just clear the current input value, such that the full list then displays.  That's the easiest method I can think of.

Would you consider a library like https://github.com/Codebrahma/React-combobox?

Gazook89 commented:
Here is a gif of a custom react component, a combobox:
![combobox](https://media.giphy.com/media/7jsfBo4uj4JZrXSGT8/giphy.gif)

The dropdown list is basically the same as the current Themes dropdown, meaning it is just a div with a collection of div options.  As such, it can be styled with css and the behavior can be changed so that it avoids overlapping the validation warning but I didn't go that far yet.

I made it because I was having trouble at first with just adding an input to the options in a satisfactory way.  Also, I have another idea in mind which would benefit from a combobox, and I thought i would make a generic component to use with that and this.  It's a little mucky right now, like a blind man dancing in a china shop.

ericscheid commented:
What happens if you type "ch" into the box, hoping to find the language code for Chinese?

Gazook89 commented:
> What happens if you type "ch" into the box, hoping to find the language code for Chinese?

It does nothing.  At least not as setup, but could be added as an option since I don't think that would be tremendously important for this language list.  But if wanting to use for another field like tags as you mentioned, then yes, that would be nifty to replicate the datalist functionality.   

If just looking to keep the datalist, but always show "all" options when initially clicking into the text input rather than the filtered list, then it's probably best to reset the input to an empty string `''` whenever you click into it.   I personally don't see any other way while still using `<datalist>`.

-----
Side note, I'm not sure someone who wants their document set to 'chinese' would type 'chinese'....I would assume they would type in chinese script?  I have no idea.

ericscheid commented:
The point of the “ch” example is that the language code for Chinese is “zh”.

Gazook89 commented:
Okay so now my custom combobox component can:

1. display a list of items in a dropdown.  The items are just divs defined like the current Languages options in the parent component and passed into this component.  
2. enable or disable auto-suggestions.  This just filters options using an `.includes()` method, which means an input of `e` returns `de, de-ch, es, en, ...`.   It could easily be changed to match only `es, en, ...` or even have that as a sub-option when enabling autoSuggest.
3. enable or disable resetting the auto-suggestions on initial click into input box.    If enabled, when you click into the input box it will show the entire option list in the dropdown.  Thereafter the dropdown will switch to autoSuggest with any text input (but not if selecting an option from the dropdown).

![combobox](https://media.giphy.com/media/xT3DHgx43PyoZwLTXD/giphy.gif)

I have this in a branch from this current PR.  Probably it could have been spun-off from *master* but I wanted to use the Language field as a test.  I'm still screwing around with it before I push it to my remote.

calculuschild commented:
I like this behavior. :+1: And with this working we should be able to apply any custom styling unlike the native datalist.

ericscheid commented:
Looks quite good.

Does the auto filter suggest `fr` is the user types `s` ?

(not that anyone would type “s” for French, but it’s a simple example that demonstrates a particular filtering behaviour)

> On 25 Nov 2022, at 2:41 pm, Gazook89 ***@***.***> wrote:
> 
> Okay so now my custom combobox component can:
> 
> display a list of items in a dropdown. The items are just divs defined like the current Languages options in the parent component and passed into this component.
> enable or disable auto-suggestions. This just filters options using an .includes() method, which means an input of e returns de, de-ch, es, en, .... It could easily be changed to match only es, en, ... or even have that as a sub-option when enabling autoSuggest.
> enable or disable resetting the auto-suggestions on initial click into input box. If enabled, when you click into the input box it will show the entire option list in the dropdown. Thereafter the dropdown will switch to autoSuggest with any text input (but not if selecting an option from the dropdown).
> 
> 
> I have this in a branch from this current PR. Probably it could have been spun-off from master but I wanted to use the Language field as a test. I'm still screwing around with it before I push it to my remote.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

Gazook89 commented:
> Does the auto filter suggest `fr` is the user types `s` ?

I just made a couple changes, so there is an option to do this.  I just published my new branch off of this one, and starting to write my PR (to merge into this, and then hopefully this into upstream/master).

Gazook89 commented:
My new PR is https://github.com/Gazook89/homebrewery/pull/1

It is a PR to merge into this branch, so I think I have to be the one to merge it but let me know if you are able to look at it and let me know.  I just didn't want to get too deep into it before merging into what is already in *this* branch.

calculuschild commented:
@Gazook89 I looked over your other PR. Looks stable enough to have you go ahead and merge it in here. We can try to push this through before it gets too much further behind Master.

Gazook89 commented:
I got my other PR merged into here with the combo-box, and I caught this up to *master*.

It's been a long time since I really looked at this, so i'm not sure where I left off-- if it was at the 'end' or somewhere in the 'middle'.  Though, it appears to work in my browser with a tiny amount of testing.

Gazook89 commented:
One issue i'm noticing is that an invalid value is still kept, even though it flashes the error.  I think likely the best route is turn invalid values to `en` as a default.

Gazook89 commented:
> As @ericscheid says, we should probably debounce the validation (different PR maybe?). Having an error show up immediately when I'm still typing is a little bit aggressive.

I spent some time trying to figure out how to do this, and couldn't get it.  I was using `_.debounce` and adding it to the `handleFieldChange` method, but it creates a new debounce method on every *input*/*keystroke* so it doesn't actually do anything worthwhile.  I also tried in a few other places, but never quite the result we want.  Some stackoverflow answers suggest it has to do with the shadow DOM and suggested react hooks or other options, but I'm just not familiar enough with the concepts that I can just put them into this PR. I think this would have to be another PR, and possibly at some point rework our input elements to work better with validation.

calculuschild commented:
Yay! Thanks @Gazook89 💯 🎉 ⭐
